### PR TITLE
[Merged by Bors] - feat(SetTheory/Ordinal/Topology): define Ordinal.IsClosed and Ordinal.IsAcc

### PR DESCRIPTION
**Pull request by YnirPaz**
Defined Ordinal.IsClosed and Ordinal.IsAcc and proved basic lemmas about them. This is closedness in the club-sense.

---

This was discussed before on #14060. `Ordinal.IsClosed` is equivalent to an appropriate topological property, and the proof will be in a later pr.

`IsClosed.sInter` and `IsClosed.iInter` can also be proven topologically once the equivalence is proven, but since they are needed for proofs unrelated to the topological equivalence I want to have them here without needing the whole topology proof.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Ordinal/Arithmetic.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Arithmetic.lean:282]: You don't need to change anything, but fyi `IsLimit` will probably be defined in terms of `IsSucc(Pre)limit` soon and this theorem will be deprecated.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:235]: I may have asked this in the previous PR but, you do need the condition `o ≠ 0` for this match the topological notion, right? Getting rid of degenerate cases should only be done for good reasons.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:263]: Not on my laptop rn, but can you golf this by doing it backwards? That is, do `apply Isclosed.sInter` and finish by `mem_range` in the next or even the same line.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:252]: Does this work?
```suggestion
    {p : Ordinal} (hp : p < o) : (U ∩ Ioo p o).Nonempty := hU.2 p hp
```

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:245]: Can you uniformize the variable names by using `S` here instead?

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:250]: Also here.

**Review by vihdzp — APPROVED** [review]
Only a few minor remarks.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (YnirPaz) [reply Mathlib/SetTheory/Ordinal/Topology.lean:235]: I added it originally so that `0` won't have to be in every club. It could also be a requirement in `IsClosed` instead, but yea then `IsAcc` won't match the topological notion. This felt like the more natural place to put it.

Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Topology.lean:235]: Alright, glad to see this is a cut and dry case.

**Comment by b-mehta** [comment]
I had a quick look through the discussion in #14060, but the conclusion there seems to be that we should define these things topologically anyway? That is, it seems more sensible to re-use IsClosed from the order topology on ordinals, and have a lemma saying it's closed in the topology sense iff `∀ p < o, (o ≠ 0 ∧ ∀ p < o, (S ∩ Ioo p o).Nonempty) → p ∈ S`.

Am I missing something?

**Comment by vihdzp** [comment]
The issue is that we want to talk about sets of ordinals being closed "below an ordinal". So the conversion between these definitions and the topological ones isn't quite so clean as it might seem. The topological equivalences are still useful to have, but I do think it's alright having two different predicates like this.

**Comment by YnirPaz** [comment]
> I had a quick look through the discussion in #14060, but the conclusion there seems to be that we should define these things topologically anyway? That is, it seems more sensible to re-use IsClosed from the order topology on ordinals, and have a lemma saying it's closed in the topology sense iff `∀ p < o, (o ≠ 0 ∧ ∀ p < o, (S ∩ Ioo p o).Nonempty) → p ∈ S`. I guess the idea is that that lemma will be pretty awkward to work with, and IsAcc is useful to have anyway... So I'm happy to have IsAcc, but I'm not convinced about Ordinal.IsClosed. What do you think @vihdzp? In any case, let's definitely add the TODO to show this is equivalent to the topology thing.

Like vi said, the topological equivalent isn't as simple as we hoped because it talks about the topology of `Iio o`, not `Ordinal`. It's `IsClosed {x : Iio o | x.1 ∈ A}`. We shouldn't write this kind of thing all the time so that's why I think it warrants a definition. I already have the topological equivalences for both somewhere and I'll PR them when this gets merged, but I'll add a TODO anyways in the meantime.

**Comment by fpvandoorn** [comment]
My two cents:
* I would also prefer to define this in terms of topological notions. In a completely different project, I've realized it is useful to have the notions `A is open in B` or `A is closed in B` in a topological space, and ideally `Ordinal.IsClosed S o` should just become `IsClosedIn S (Iio o)`, and something similar for accumulation points.
* I am happy to quickly merge API about `IsClosedIn`/`IsOpenIn` for topological spaces (maybe after a brief message on Zulip if there are any objections to this idea). 
* We could merge this in the meantime, with the idea (and an explicit comment) that this should be refactored later. If we do that, the name `IsClosed` should probably be `IsClosedBelow` or something.

**Comment by vihdzp** [comment]
I like this `IsOpenIn` idea. I'm not really sure how its API should look though, do we just translate the basic results on open/closed sets to this setting and leave it at that?

**Comment by fpvandoorn** [comment]
We can start small: just have all the lemmas how `IsOpenIn`/`IsClosedIn` interacts with intersections, unions, `univ` and `IsOpen`/`IsClosed` (e.g. if `B` is closed then `IsClosedIn A B \iff IsClosed (A \inter B)`). We can increase the API as we use it more.
If you want to write the basic API for this, that would be great!

**Comment by urkud** [comment]
You can use https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Notation.html#Set.Notation.%C2%ABterm_%E2%86%93%E2%88%A9_%C2%BB to define `IsOpenIn`/`IsClosedIn`.

**Comment by vihdzp** [comment]
I've opened that PR in #17854. For now I've only added relative open/closed sets; if there's no objections I'll add `AccPtIn` as well in a follow-up.

**Comment by YnirPaz** [comment]
> My two cents:
> 
> * I would also prefer to define this in terms of topological notions. In a completely different project, I've realized it is useful to have the notions `A is open in B` or `A is closed in B` in a topological space, and ideally `Ordinal.IsClosed S o` should just become `IsClosedIn S (Iio o)`, and something similar for accumulation points.
> * I am happy to quickly merge API about `IsClosedIn`/`IsOpenIn` for topological spaces (maybe after a brief message on Zulip if there are any objections to this idea).
> * We could merge this in the meantime, with the idea (and an explicit comment) that this should be refactored later. If we do that, the name `IsClosed` should probably be `IsClosedBelow` or something.

I'm renaming and adding the TODO. Did you end up opening a zulip discussion about this?

**Comment by vihdzp** [comment]
Yes, and now it seems that it's contentious whether we even want relative open/closed sets given the `∩↓` is available. At this point I'd rather just see this PR merged rather than keep stalling.

**Comment by fpvandoorn** [comment]
I disagree. I think this PR has to change to `IsOpenIn s t` or `IsOpen (s ↓∩ t)`, and only the former is currently stuck by Zulip. I'm happy to merge it if you adapt the code to use the latter option.

About the accumulation point, is that the same as [AccPt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Defs/Filter.html#AccPt) (without relativization), since an ordinal cannot be approached from above? If so, please also use that

**Comment by fpvandoorn** [comment]
Ah, I see in the Zulip thread that `IsClosedBelow` is `IsClosed ((S ∩ Iio o) ∪ {o})`, so it's not even a special case of `IsClosedIn`? My bad then, then that API wouldn't help here.

Could you add the equivalence between `IsClosedBelow` and `IsClosed` of that complicated set? Or if that's hard, we can merge it as-is.

**Comment by YnirPaz** [comment]
> Ah, I see in the Zulip thread that `IsClosedBelow` is `IsClosed ((S ∩ Iio o) ∪ {o})`, so it's not even a special case of `IsClosedIn`? My bad then, then that API wouldn't help here.
> 
> Could you add the equivalence between `IsClosedBelow` and `IsClosed` of that complicated set? Or if that's hard, we can merge it as-is.

It is a special case. `IsClosedBelow` is equivalent to `IsClosed (Iio o ↓∩ S)`, `((S ∩ Iio o) ∪ {o})` is just adding all the possible acc points of `S ∩ Iio o` to it (only 1) so that asking if it's closed is the same as asking if it's closed inside `Iio o`.

I'm currently changing the definition to `IsClosed (Iio o ↓∩ S)` and adding a theorem characterizing it with the current def, I think that's what people want? tbh I still think it's silly not using the original definitions and having the topological ones as theorems (to me it's like defining limit ordinals as accumulation points of `Ord`, which is obviously not done), but this was argued over a lot so now I'm just going along with whatever gets this merged.

**Comment by vihdzp** [comment]
> to me it's like defining limit ordinals as accumulation points of `Ord`, which is obviously not done

Note that as part of my planned ordinal refactors (#17033), `IsLimit` will get replaced with the more general [`IsSuccLimit`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/SuccPred/Limit.html#Order.IsSuccLimit). Part of what makes the current definition annoying is that the API for it had to be duplicated for ordinals and cardinals. Redefining `IsLimit` as an accumulation point would have been another way to solve that problem :P

File: Mathlib/Topology/Basic.lean
Comment (vihdzp) [inline Mathlib/Topology/Basic.lean]: Note that `h` is equivalent to `x ∈ closure C` by `mem_closure_iff_clusterPt`. Also, isn't this an iff?

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:373]: I think this should work
```suggestion
alias ⟨IsClosedBelow.forall_lt, _⟩ := isClosedBelow_iff
```

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:357]: Iffs generally have their arguments implicit.
```suggestion
theorem isClosedBelow_iff {S : Set Ordinal} {o : Ordinal} : IsClosedBelow S o ↔
```

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (YnirPaz) [reply Mathlib/SetTheory/Ordinal/Topology.lean:373]: It does but then you need an extra parentheses around `.forall_lt`, ie `(h C CmemS).forall_lt p plto ...` -> `((h C CmemS).forall_lt) p plto ...`

File: Mathlib/Topology/Basic.lean
Comment (YnirPaz) [reply Mathlib/Topology/Basic.lean]: I can make it into an iff but will the other direction ever be used? We have both `Or` cases as theorems already.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Topology.lean]: That doesn't seem right, function application doesn't need parentheses.

File: Mathlib/Topology/Basic.lean
Comment (vihdzp) [reply Mathlib/Topology/Basic.lean]: Well, the other thing nice about iffs is that you can use them in `rw`. That seems at least somewhat plausible.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (YnirPaz) [reply Mathlib/SetTheory/Ordinal/Topology.lean]: Alright, I think i figured it out. Creating the alias automatically turned some arguments from implicit to explicit, arguments like `S` that are part of the definition of `IsClosedBelow`. So then `(h C CmemS).forall_lt p` didn't typecheck because `(h C CmemS).forall_lt` expected the now explicit arguments `S` and `o`. But apparantly surrounding an expressing like that with parentheses forces lean to fill in these inputs, so `((h C CmemS).forall_lt) = (h C CmemS).forall_lt C o`.

Also I'm now seeing that this problem just went away hahaha. I must have copied something wrong and somehow made the alias have explicit inputs or something.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (fpvandoorn) [inline Mathlib/SetTheory/Ordinal/Topology.lean:296]:
```suggestion
theorem IsAcc.mono {o : Ordinal} {S T : Set Ordinal} (h : S ⊆ T) (ho : o.IsAcc S) :
```
I think `mono` is a much more common name in Mathlib.

**Review by fpvandoorn — APPROVED** [review]
This looks good to me now, up to some minor things.

I agree that `clusterPt_principal` is nicer as an `iff`.

File: Mathlib/SetTheory/Ordinal/Topology.lean
Comment (vihdzp) [inline Mathlib/SetTheory/Ordinal/Topology.lean:304]: I don't think this should hold up this PR since this is an issue with a lot of the other theorems in the file but, this should hold for any linear `SuccOrder`, right? If so it'd be nice to add a TODO.

**Comment by vihdzp** [comment]
Just reasserting my LGTM. I'm excited to see club sets in the near future!

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**
